File: Algorithm/python_source/charpter16_slide/level2/length_of_longest_substring_k_distinct.py
# coding=utf-8
from collections import defaultdict


class LengthOfLongestSubstringKDistinct:
    def lengthOfLongestSubstringKDistinct(self, s, k):
        n = len(s)
        if k == 0 or n == 0:
            return 0
        left, right = 0, 0
        hashmap = defaultdict()
        max_len = 1
        while right < n:
            hashmap[s[right]] = right
            right += 1
            if len(hashmap) == k + 1:
                del_idx = min(hashmap.values())
                del hashmap[s[del_idx]]
                left = del_idx + 1

            max_len = max(max_len, right - left)
        return max_len


if __name__ == '__main__':
    s = "eceba"
    k = 2
    lengthOfLongestSubstringKDistinct = LengthOfLongestSubstringKDistinct()
    print lengthOfLongestSubstringKDistinct.lengthOfLongestSubstringKDistinct(s, k)

File: Algorithm/Code_Materials/Java/src/main/java/com/yugutou/charpter16_slide/level2/LengthOfLongestSubstringKDistinct.java
package com.yugutou.charpter16_slide.level2;

import java.util.Collections;
import java.util.HashMap;

public class LengthOfLongestSubstringKDistinct {
    public static void main(String[] args) {
        String s = "eceba";
        int k = 2;
        System.out.println(lengthOfLongestSubstringKDistinct(s, k));
    }

    public static int lengthOfLongestSubstringKDistinct(String s, int k) {
        if (s.length() < k + 1) {
            return s.length();
        }

        int left = 0, right = 0;
        HashMap<Character, Integer> hashmap = new HashMap<>();
        int maxLen = k;

        while (right < s.length()) {

            if (hashmap.size() < k + 1)
                hashmap.put(s.charAt(right), right++);

            // 如果大小达到了k个
            if (hashmap.size() == k + 1) {
                //
                int del_idx = Collections.min(hashmap.values());
                hashmap.remove(s.charAt(del_idx));
                // 窗口left的新位置
                left = del_idx + 1;
            }

            maxLen = Math.max(maxLen, right - left);
        }
        return maxLen;
    }
}
